Modules/Scripted/DICOMLib/SubjectHierarchyPlugins/qSlicerSubjectHierarchyDICOMPlugin.cxx

```diff
@@ -429,6 +433,99 @@ void qSlicerSubjectHierarchyDICOMPlugin::openDICOMExportDialog()
     return;
     }
 
+  // Check whether there is a study and patient for the exported item
+  bool validSelection = false;
+  if (shNode->IsItemLevel(currentItemID, vtkMRMLSubjectHierarchyConstants::GetDICOMLevelStudy()))
+    {
+    if (shNode->IsItemLevel(shNode->GetItemParent(currentItemID), vtkMRMLSubjectHierarchyConstants::GetDICOMLevelPatient()))
+      {
+      // Valid if current item is study and has a patient parent
+      validSelection = true;
+      }
+    }
+  else // Item belongs to a data node (i.e. series)
+    {
+    vtkIdType studyItemID = shNode->GetItemParent(currentItemID);
+    if ( studyItemID && shNode->IsItemLevel(studyItemID, vtkMRMLSubjectHierarchyConstants::GetDICOMLevelStudy())
+      && shNode->IsItemLevel(shNode->GetItemParent(studyItemID), vtkMRMLSubjectHierarchyConstants::GetDICOMLevelPatient()) )
+      {
+      // Valid if current item is a series and has a study parent which has a patient parent
+      validSelection = true;
+      }
+    }
+  if (!validSelection)
+    {
+    QString message = QString("Data to export need to be under a study item with a parent patient.\n"
+                              "Default patient and study will be created and the selected data and its related datasets "
+                              "will be moved in it for export.\n\n"
+                              "If you'd like to create the hierarchy manually, please click Cancel, "
+                              "then create a patient by right-clicking the empty area in Subject hierarchy in the Data module "
+                              "and choosing 'Create new subject'. Study can be similarly created under the patient. The data "
+                              "to export can be drag&dropped under the study.");
+    QMessageBox::StandardButton answer =
+      QMessageBox::question(NULL, tr("Create new patient and study for DICOM export?"), message,
+      QMessageBox::Ok | QMessageBox::Cancel, QMessageBox::Ok);
+    if (answer == QMessageBox::Ok)
+      {
+      // Generate new patient name
+      std::string patientName = vtkMRMLSubjectHierarchyConstants::GetSubjectHierarchyNewItemNamePrefix()
+        + vtkMRMLSubjectHierarchyConstants::GetDICOMLevelPatient();
+      patientName = shNode->GenerateUniqueItemName(patientName);
+      // Create parent patient
+      vtkIdType patientItemID = shNode->CreateSubjectItem(shNode->GetSceneItemID(), patientName);
+
+      if (shNode->IsItemLevel(currentItemID, vtkMRMLSubjectHierarchyConstants::GetDICOMLevelStudy()))
+        {
+        // Move existing study under new patient
+        shNode->SetItemParent(currentItemID, patientItemID);
+        }
+      else
+        {
+        // Collect referenced items for the selected data item
+        // DICOM references
+        std::vector<vtkIdType> referencedItems = shNode->GetItemsReferencedFromItemByDICOM(currentItemID);
+        // MRML references available in SH
```

Comment:
Do we only add the nodes that are referenced directly? Shouldn't we add all referenced nodes recursively? (you can get that list of nodes using vtkMRMLScene::GetReferencedSubScene)